test/NuGet.Core.Tests/NuGet.PackageManagement.Test/BuildIntegration/DependencyGraphRestoreUtilityTests.cs

```diff
@@ -140,5 +144,215 @@ public async Task RestoreAsync_WithMinimalProjectAndAdditionalErrorMessage_Write
+            RestoreSummary restoreSummary = result[0];
+            restoreSummary.Success.Should().BeTrue();
+            restoreSummary.NoOpRestore.Should().BeFalse();
+            var assetsFilePath = Path.Combine(packageSpec.RestoreMetadata.OutputPath, LockFileFormat.AssetsFileName);
+            File.Exists(assetsFilePath).Should().BeTrue(because: $"{assetsFilePath}");
+            progressReporter.StartProjectUpdateQueue.Count.Should().Be(1);
+            progressReporter.EndProjectUpdateQueue.Count.Should().Be(1);
+            var startProjectUpdateData = progressReporter.StartProjectUpdateQueue.Dequeue();
+            var endProjectUpdateData = progressReporter.EndProjectUpdateQueue.Dequeue();
+
+            startProjectUpdateData.Should().Be(endProjectUpdateData);
+
+            startProjectUpdateData.Item1.Should().Be(packageSpec.FilePath);
+            startProjectUpdateData.Item2.Should().HaveCount(3);
+
+            var propsFile = BuildAssetsUtils.GetMSBuildFilePath(packageSpec, BuildAssetsUtils.PropsExtension);
+            var targetsFile = BuildAssetsUtils.GetMSBuildFilePath(packageSpec, BuildAssetsUtils.TargetsExtension);
+            startProjectUpdateData.Item2.Should().Contain(assetsFilePath);
+            startProjectUpdateData.Item2.Should().Contain(propsFile);
+            startProjectUpdateData.Item2.Should().Contain(targetsFile);
+        }
+
+        [Fact]
+        public async Task RestoreAsync_WithProgressReporter_WhenProjectNoOps_ProgressIsNotReported()
+        {
+            // Arrange
+            using var pathContext = new SimpleTestPathContext();
+            var settings = Settings.LoadDefaultSettings(pathContext.SolutionRoot);
+            var packageSpec = ProjectTestHelpers.GetPackageSpec(settings, projectName: "projectName", rootPath: pathContext.SolutionRoot);
+            var dgSpec = ProjectTestHelpers.GetDGSpecFromPackageSpecs(packageSpec);
+            var progressReporter = new RestoreProgressReporter();
+
+            // Pre-Conditions
+            IReadOnlyList<RestoreSummary> result = await DependencyGraphRestoreUtility.RestoreAsync(
+                dgSpec,
+                new DependencyGraphCacheContext(),
+                new RestoreCommandProvidersCache(),
+                cacheContextModifier: _ => { },
+                sources: new SourceRepository[0],
+                parentId: Guid.Empty,
+                forceRestore: false,
+                isRestoreOriginalAction: true,
+                additionalMessages: null,
+                progressReporter: progressReporter,
+                new TestLogger(),
+                CancellationToken.None);
+
+            // Pre-Conditions
+            result.Should().HaveCount(1);
+            RestoreSummary restoreSummary = result[0];
+            restoreSummary.Success.Should().BeTrue();
+            restoreSummary.NoOpRestore.Should().BeFalse();
+
+            progressReporter.StartProjectUpdateQueue.Count.Should().Be(1);
```

Comment:
Some people might not like the syntax, but I think using Moq, you can write fewer lines of code and avoid the test class implementing the interface.

`var mock = Mock<IRestoreProgressReporter>()` will allow you to `mock.Validate(r => r.StartProjectUpdate(It.IsAny<string>(), It.IsAny<string>()), Times.One)` or `Times.Zero`. Instead of multiple dequeue lines, just create a new `Mock<T>` instance, as I don't think it provides the ability to clear its invocation list.

Reply:
I shifted it mock, but for the 3rd test, it's actually *way* longer than the original version.